### ISSUE TITLE
'acf_qtranslate_acf_5_image' does not have a method 'wp_prepare_attachment_for_js'

### ISSUE THREAD

**Issue by fabienlege** [qtranslate/qtranslate-xt#718]
Wordpress media library not working with build-in acf image field.
When click on "select image", media library open, but show no image nor error.

In php log, we can see this error : `PHP Warning:  call_user_func_array() expects parameter 1 to be a valid callback, class 'acf_qtranslate_acf_5_image' does not have a method 'wp_prepare_attachment_for_js' in /var/www/html/cauterets/wp-includes/class-wp-hook.php on line 286`

I use the last version of your plugin and acf

**Comment by Tusko** [comment]
+1

**Comment by WIDESRL** [comment]
+1

**Comment by fminov** [comment]
Hi,

Exact same problem here, site with 1600+ images, seems like an ajax call is made for each image before the result is displayed in media library. So 1600+ ajax calls before seeing images in the library.
Same error in log Fabien posted, repeated for each ajax call (fast growing log !).

Thanks for your help, please tag this as bug.

**Comment by Tusko** [comment]
@fminov Media library by default doesn't work.

**Comment by fminov** [comment]
@Tusko hi, indeed, same problem with Real Media Library enabled or with stock Media Library

**Comment by herrvigg** [comment]
> Wordpress media library not working with build-in acf image field.

hello,
this looks related to ACF, i don't see why this would be a problem with qTranslate especially since you are using a standard ACF image field! I could understand if you were using a qTranslate image field but not really here. What happens if you disable qTranslate-XT?

**Comment by gab1982** [comment]
> 
> 
> > Wordpress media library not working with build-in acf image field.
> 
> hello,
> this looks related to ACF, i don't see why this would be a problem with qTranslate especially since you are using a standard ACF image field! I could understand if you were using a qTranslate image field but not really here. What happens if you disable qTranslate-XT?

Yes but it is ok when qTranslate-XT is disabled as I mentioned in issue 719, so somehow they must be related, I was using the ACF gallery, going back to the previous version of ACF (5.8.2) fixed the issue for me.

**Comment by herrvigg** [comment]
Mmmm that sounds strange but let's see.

1) Does this happen if you disable ACF and keep qTranslate-xt?

2) With both ACF and qTranslate-XT enabled, can you try just to disable manually the ACF module in qTranslate-XT? Either rename `qtranslate-xt/modules/acf` to something else (e.g. `acf-disabled`), or comment the acf entry in `modules/qtx_modules_handler.php`. This will just disable temporarily the specific integration part with ACF (just work on test posts under that time).

**Comment by gab1982** [comment]
> Mmmm that sounds strange but let's see.
> 
> 1. Does this happen if you disable ACF and keep qTranslate-xt?
> 2. With both ACF and qTranslate-XT enabled, can you try just to disable manually the ACF module in qTranslate-XT? Either rename `qtranslate-xt/modules/acf` to something else (e.g. `acf-disabled`), or comment the acf entry in `modules/qtx_modules_handler.php`. This will just disable temporarily the specific integration part with ACF (just work on test posts under that time).

1. No it seems ok if I disable ACF
2. If I rename the acf folder to something else, it seems to fix the issue

thanks

**Comment by herrvigg** [comment]
OK i found it

https://github.com/qtranslate/qtranslate-xt/blob/4ac406e999bb8dae0a87e3be5c8789d38d577f64/modules/acf/src/acf_5/fields/image.php#L75

As a quick workaround fix you could just comment this line. For the correct fix one should look at the new `acf_image_field`. This means they have changed it in the last version (5.8.3) and this `wp_prepare_attachment_for_js` function has disappeared. If it works with 5.8.2 this is really bad from ACF as a patch release should not break the current interface!

Btw where did you get ACF 5.8.3?  Maybe it's only for the PRO version. For the free it's still [5.8.2 even on github](https://github.com/AdvancedCustomFields/acf/releases).

**Comment by gab1982** [comment]
> OK i found it
> 
> https://github.com/qtranslate/qtranslate-xt/blob/4ac406e999bb8dae0a87e3be5c8789d38d577f64/modules/acf/src/acf_5/fields/image.php#L75
> 
> As a quick workaround fix you could just comment this line. For the correct fix one should look at the new `acf_image_field`. This means they have changed it in the last version (5.8.3) and this `wp_prepare_attachment_for_js` function has disappeared. If it works with 5.8.2 this is really bad from ACF as a patch release should not break the current interface!
> 
> Btw where did you get ACF 5.8.3? Maybe it's only for the PRO version. For the free it's still [5.8.2 even on github](https://github.com/AdvancedCustomFields/acf/releases).

Thanks for the quick reply and quick fix, yes its not good, I have the pro version, hopefully they release a fix soon.

**Comment by fminov** [comment]
@herrvigg Hi, thanks a lot for the quick fix, it works perfectly !

**Comment by herrvigg** [comment]
Nice to hear! However i'm not sure about the impact, disabling this may have some side effects for that particular field.

Could someone having 5.8.3 send the new file holding the `acf_image_field` class? So we can see what they have done there. Some of the PRO users should also open a ticket on their side at least to tell them they broke the interface.

**Comment by fminov** [comment]
Is there any simple way to contact you privately for this ?

Note : previously I tried ACF 5.8.2 Pro and it didn't work any better.

**Comment by gab1982** [comment]
> acf_image_field

I searched the plugin for this but no where to be found.

**Comment by herrvigg** [comment]
> Is there any simple way to contact you privately for this ?

`<my-nickname> at gmail.com`
 
> Note : previously I tried ACF 5.8.2 Pro and it didn't work any better.

This sounds weird, in contradiction with what @gab1982 mentioned earlier :-/
Also, see the code of ACF 5.8.2 below. the function is there and i don't see how it would be different in the PRO version but please check below.

> I searched the plugin for this but no where to be found.

We derive from `acf_image_field` so this has to be there otherwise you'd have an error. But maybe the have done something special with the Pro release.

Here is the class and the concerned function in 5.8.2 Free:  `acf/includes/fields/class-acf-field-image.php`

https://github.com/AdvancedCustomFields/acf/blob/f745d197118d1396a45084ddad6299abdf78d72d/includes/fields/class-acf-field-image.php#L396

Now to come up with a fix one should look at the diff 5.8.3 PRO vs 5.8.2 PRO.
And maybe the diff 5.8.2 Free vs 5.8.2 PRO (just about this function).

**Comment by herrvigg** [comment]
All right, tthey simply removed the function from 5.8.3 and it's not documented in the release notes. But it looked like a hacky hook in the first place:

```
*  wp_prepare_attachment_for_js
	*
	*  this filter allows ACF to add in extra data to an attachment JS object
	*  This sneaky hook adds the missing sizes to each attachment in the 3.5 uploader. 
	*  It would be a lot easier to add all the sizes to the 'image_size_names_choose' filter but 
	*  then it will show up on the normal the_content editor
```	

So we can certainly assume it is safe to ignore it. Maybe they fixed something in WP core that allows to avoid this and i don't expect it to see it back again. I'll find a way to have a better check in QT-XT. For now the best temporary fix is simply to comment the line in the ACF module as described earlier.

**Comment by herrvigg** [comment]
Fix merged in master.

**Comment by herrvigg** [comment]
@fminov it is really surprising it didn't work when you moved back to ACF 5.8.2 because the original function is there. Maybe you had some cache working or something else happened when switching the versions.

**Comment by fminov** [comment]
@herrvigg thanks a lot for your research !
Maybe I had the problem with 5.8.2 because of cache yes, it's possible.

**Comment by GersonM** [comment]
Hi, 

there are no update yet but the fix works fine.

I found the archive here 
wp-content/plugins/acf-qtranslate/src/acf_5/fields

Regards!.

**Comment by netgfx** [comment]
Hello is this fixed on some latest version? I have `5.8.11` (pro) and it is still affecting media library.

**Comment by herrvigg** [comment]
@netgfx this has been fixed long ago. What qTranslate-XT are you using?

**Comment by netgfx** [comment]
@herrvigg Well that's the thing I have qtranslate-X (the site already had it installed when I picked it up), I know its long abandoned, do you think it is fairly save to replace it with qtranslate-xt or would there be lots of issues?
I managed to fix it by commenting the `acf_image_field` but its a hacky solution.

**Comment by herrvigg** [comment]
@netgfx why do you still use qTranslate-X? There is no reason using an obsolete plugin. If you read the main README you will find the answer to your questions.

**Comment by yokotsumo** [comment]
ACF: 5.8.12
Qtranslate-XT : 3.6.0

I have the same problem and the above solution is not working.

**Comment by herrvigg** [comment]
You have all the history in this thread. Fixed in 3.6.2 almost one year ago... why are you using such an old version? Last release is 3.8.1.

**Comment by solvewp** [comment]
ACF: 5.8.12
qTranslate-XT: 3.8.1

Same issue exist!
